server/src/main/java/org/apache/druid/segment/realtime/appenderator/SinkQuerySegmentWalker.java

```diff
@@ -196,13 +200,26 @@ public <T> QueryRunner<T> getQueryRunnerForSegments(final Query<T> query, final
     final LinkedHashMap<SegmentDescriptor, List<QueryRunner<T>>> allRunners = new LinkedHashMap<>();
 
     try {
-      for (final SegmentDescriptor newDescriptor : specs) {
-        final SegmentDescriptor descriptor = newIdToBasePendingSegment.getOrDefault(newDescriptor, newDescriptor);
-        final PartitionChunk<Sink> chunk = sinkTimeline.findChunk(
+      for (final SegmentDescriptor descriptor : specs) {
+        final PartitionChunk<SegmentDescriptorOvershadowable> descriptorChunk = upgradedSegmentsTimeline.findChunk(
             descriptor.getInterval(),
             descriptor.getVersion(),
             descriptor.getPartitionNumber()
         );
+        final SegmentDescriptor baseDescriptor;
+        if (descriptorChunk != null) {
+          baseDescriptor = newIdToBasePendingSegment.getOrDefault(
```

Comment:
If the descriptor was found in `upgradedSegmentsTimeline`, but not found in `newIdToBasePendingSegment`, then something went wrong, didn't it? Unless I am missing something, it seems like it would make more sense to throw a defensive exception than to try to look up the original `descriptor`.

Reply:
I've simply remove the orDefault part